fc/parsing/actions.py

```diff
@@ -1413,63 +1630,53 @@ def _handle_clamping(self):
                 self.model.remove_equation(defn)
             self.model.add_equation(new_defn)
 
-    def _annotate_state_variables(self):
-        """Annotate all state variables with the 'magic' oxmeta:state_variable term."""
-        state_annotation = create_rdf_node((OXMETA_NS, 'state_variable'))
-        self.vector_orderings[state_annotation] = {}
-        for i, state_var in enumerate(self.model.get_state_variables()):
-            self.model.add_cmeta_id(state_var)
-            self.model.rdf.add((state_var.rdf_identity, PRED_IS_VERSION_OF, state_annotation))
-            self.vector_orderings[state_annotation][state_var.rdf_identity] = i
-
-    def _convert_output_units(self):
-        """Convert units for all outputs if needed.
-
-        :return: The set of variables appearing in outputs, either directly or as part of a vector, in the desired
-            units.
+    def _gather_state_variables(self, original_ordering):
+        """
+        Gather all variables annotated as states, and create a vector ordering. If required, update the ProtocolVariable
+        for the oxmeta:state_variable annotation.
+
+        :param original_ordering: An list containing the rdf identities of the variables as they originally appeared
+            (before any model manipulation).
         """
-        output_variables = set()
-        for output in self.outputs:
-            variables = get_variables_transitively(self.model, output.rdf_term)
-            if output.units is not None:
-                desired_units = self.units.get_unit(output.units)
-                for i, variable in enumerate(variables):
-                    variables[i] = self.model.convert_variable(
-                        variable, desired_units, DataDirectionFlow.OUTPUT)
-            output_variables.update(variables)
-        return output_variables
-
-    def _list_parameters(self):
-        """Populates the list of model parameters (inputs with a constant RHS)."""
-
-        for var in self.inputs:
+        # Gather state variables
+        states = get_variables_that_are_version_of(self.model, STATE_ANNOTATION)
+
+        # Create and store ordering, preserving original ordering as much as possible
+        order = []
+        todo = set([var.rdf_identity for var in states])
+        for rdf_identity in original_ordering:
             try:
-                variable = self.model.get_variable_by_ontology_term(var.rdf_term)
+                todo.remove(rdf_identity)
             except KeyError:
-                # Skip optional variables
                 continue
+            order.append(rdf_identity)
+        order.extend(todo)
```

Comment:
So we'd need to ensure we sort `todo` into a predictable order here; perhaps just a string sort on the contents? So `order.extend(sorted(todo))`?